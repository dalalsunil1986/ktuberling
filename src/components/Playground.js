/* eslint-disable class-methods-use-this */
import React from 'react';
import PropTypes from 'prop-types';

import PlaygroundJs from '../playground/playground';


export default class Playground extends React.Component {
  static propTypes = { ground: PropTypes.object }
  static defaultProps = { ground: undefined }


  constructor(props) {
    super(props);
    this.state = { loading: true };
  }


  _setPlayground = (playground) => {
    if (!playground) {
      return;
    }

    this._playground.setPlayground(playground);
  }


  _setRef = (ref) => {
    this._playgroundRef = ref;
  }


  componentWillReceiveProps(nextProps) {
    if (this.props.ground !== nextProps.ground) {
      this._setPlayground(nextProps.ground);
    }
  }


  shouldComponentUpdate() {
    return false;
  }


  componentDidMount() {
    this._playground = new PlaygroundJs(this._playgroundRef);
    this._setPlayground(this.props.ground);
  }


  componentWillUnmount() {
    this._playground.remove();
    this._playground = undefined;
  }


  render() {
    return (
      <div ref={this._setRef} className="playground" />
    );
  }
}
