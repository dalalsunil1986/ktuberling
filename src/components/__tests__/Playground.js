import React from 'react';
import renderer from 'react-test-renderer';

import Playground from '../Playground';
import PlaygroundJs from '../../playground/playground';

jest.mock('../../playground/playground', () => jest.fn());


describe('Playground', () => {
  const ground = {
    name: 'ground',
    file: 'ground.svg',
    localization: {
      bs: 'bs',
      ca: 'bs',
    },
    bgcolor: 'bgcolor',
    items: [
      {
        name: 'item 1',
        scale: '8',
        sound: 'item1',
      },
    ],
  };


  beforeEach(() => {
    PlaygroundJs.mockReset();
  });


  it('render', () => {
    const remove = jest.fn();
    const setPlayground = jest.fn();
    PlaygroundJs.mockImplementation(() => ({ remove, setPlayground }));

    const tree = renderer.create(<Playground ground={ground} />);

    expect(tree).toMatchSnapshot();
    expect(PlaygroundJs).toHaveBeenCalledTimes(1);
    expect(PlaygroundJs.mock.calls).toMatchSnapshot();
    expect(setPlayground).toHaveBeenCalledTimes(1);
    expect(setPlayground.mock.calls).toMatchSnapshot();

    tree.unmount();

    expect(remove).toHaveBeenCalledTimes(1);
    expect(remove.mock.calls).toMatchSnapshot();
  });


  it('render empty', () => {
    const remove = jest.fn();
    const setPlayground = jest.fn();
    PlaygroundJs.mockImplementation(() => ({ remove, setPlayground }));

    const tree = renderer.create(<Playground ground={undefined} />);

    expect(tree).toMatchSnapshot();
    expect(PlaygroundJs).toHaveBeenCalledTimes(1);
    expect(PlaygroundJs.mock.calls).toMatchSnapshot();
    expect(setPlayground).toHaveBeenCalledTimes(0);

    tree.unmount();

    expect(remove).toHaveBeenCalledTimes(1);
    expect(remove.mock.calls).toMatchSnapshot();
  });


  it('update playground', () => {
    const groundBefore = {
      ...ground,
      name: 'groundBefore',
    };
    const groundAfter = {
      ...ground,
      name: 'groundAfter',
    };
    const setPlayground = jest.fn();
    PlaygroundJs.mockImplementation(() => ({ setPlayground }));

    const tree = renderer.create(<Playground ground={groundBefore} />);

    expect(tree).toMatchSnapshot('update playground tree');
    expect(PlaygroundJs).toHaveBeenCalledTimes(1);
    expect(PlaygroundJs.mock.calls).toMatchSnapshot();
    expect(setPlayground).toHaveBeenCalledTimes(1);
    expect(setPlayground.mock.calls).toMatchSnapshot();

    // same ground if property not updated
    tree.update(<Playground ground={groundBefore} />);
    expect(PlaygroundJs).toHaveBeenCalledTimes(1);
    expect(setPlayground).toHaveBeenCalledTimes(1);

    // new ground
    tree.update(<Playground ground={groundAfter} />);

    expect(tree).toMatchSnapshot('update playground tree');
    expect(PlaygroundJs).toHaveBeenCalledTimes(1);
  });
});
