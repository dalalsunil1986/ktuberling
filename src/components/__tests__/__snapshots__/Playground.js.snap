// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playground render 1`] = `
<div
  className="playground"
/>
`;

exports[`Playground render 2`] = `
Array [
  Array [
    null,
  ],
]
`;

exports[`Playground render 3`] = `
Array [
  Array [
    Object {
      "bgcolor": "bgcolor",
      "file": "ground.svg",
      "items": Array [
        Object {
          "name": "item 1",
          "scale": "8",
          "sound": "item1",
        },
      ],
      "localization": Object {
        "bs": "bs",
        "ca": "bs",
      },
      "name": "ground",
    },
  ],
]
`;

exports[`Playground render 4`] = `
Array [
  Array [],
]
`;

exports[`Playground render empty 1`] = `
<div
  className="playground"
/>
`;

exports[`Playground render empty 2`] = `
Array [
  Array [
    null,
  ],
]
`;

exports[`Playground render empty 3`] = `
Array [
  Array [],
]
`;

exports[`Playground update playground 1`] = `
Array [
  Array [
    null,
  ],
]
`;

exports[`Playground update playground 2`] = `
Array [
  Array [
    Object {
      "bgcolor": "bgcolor",
      "file": "ground.svg",
      "items": Array [
        Object {
          "name": "item 1",
          "scale": "8",
          "sound": "item1",
        },
      ],
      "localization": Object {
        "bs": "bs",
        "ca": "bs",
      },
      "name": "groundBefore",
    },
  ],
]
`;

exports[`Playground update playground: update playground tree 1`] = `
<div
  className="playground"
/>
`;

exports[`Playground update playground: update playground tree 2`] = `
<div
  className="playground"
/>
`;
